### ISSUE TITLE
Add type coercion/type casting comparisons example

### ISSUE THREAD

**Issue by ostrolucky** [Girgias/unify-typing-modes-rfc#2]
There is a nice discussion about it in symfony slack here https://symfony-devs.slack.com/archives/C8WHX21K7/p1682583466502569

I think your README could benefit from showing the examples of type coercion https://3v4l.org/NERuj

Some quote from the discussion I liked
> The issue with strict types is that it forces you to use type casting to replace places where non-strict mode will do type coercion and that's where they become less type safe. Strict_types is all about forbidding type coercion (which then force you to use type casting instead). That's why Symfony does not use strict types (and avoids type casting as much as possible)

**Comment by Girgias** [comment]
I tried opening the Slack discussion, but it doesn't want me to access it, and I frankly don't understand Slack and all its weirdness...

In any case, I need to update the document as some stuff has changed in the meantime, but adding examples is a good idea.

**Comment by stof** [comment]
That discussion is gone now anyway (due to the Slack retention policy for accounts on the free plan).

This discussion was about some people asking why Symfony does not use strict_types in its codebase. This mostly boils down to what the readme mentions as ["too strict may lead to too lax"](https://github.com/Girgias/unify-typing-modes-rfc?tab=readme-ov-file#2-too-strict-may-lead-to-too-lax).

The 3v4l link was me showing the effect of this for integers (I did not spend time doing similar comparison for other types, either because they have less weird cases or because coercion rules are still weird for them as well) by comparing the effect of using an explicit `(int)` vs letting PHP perform type coercion. This was about proving that the type coercion actually report an error in *more* cases (and a deprecation if some of the cases where we would want an error so that will improve even more in PHP 9). I did that because I figured out that other people in the discussion were not aware of the very lax behavior of the explicit cast (and to be honest, even I was surprised by the case of arrays not even triggering a warning)

**Comment by Girgias** [comment]
Right, for comparisons I've been working on a draft RFC that you can find on an other git repo: https://github.com/Girgias/php-rfcs/blob/master/comparison-equality-semantics.md

Because I do agree the type juggling in the comparative context is quite bonkers, but `strict_types` does not affect comparisons, which is why I don't think it's useful to add such examples to this RFC.

**Comment by stof** [comment]
@Girgias AFAIU, this issue was not about adding examples for the type juggling in comparative context but about adding a comparison table between the type juggling in the function context and explicit type casts (which are often used in `strict_types` to "fix" type errors once type-juggling is gone)

**Comment by Girgias** [comment]
Ahhhhhhhhh, okay yeah that is a somewhat sensible idea.

Will keep it in mind when I go back to edit this document!

**Comment by stof** [comment]
For reference, I updated my 3v4l snippet to account for more types:
- I added `false` (and renamed the `bool` case to `true`) as `true` and `false` often have different behavior
- I added a case for `GMP` as internal objects overloading some behavior can have a different behavior in type juggling (I could not add the weird case of the Tidy object as ext-tidy is not available on 3v4l.org)
- I added a case for `resource` which I forgot in the past

The new snippet at https://3v4l.org/f1ohY should be covering all cases coercing/casting to `int` now.
